# Assignment 2 - Correlation between the ranking of Human Development Index (HDI) values and the values of the factors that make up the HDI

## Introduction

The Human Development Index (HDI) ranks countries based on their standard of living. The purpose of the index is to assess the standard of living within most of the sovereign nations and territories around the world. While the index is far from perfect, it is useful as a general overview of the living standards one could expect should one be born or live in any of the nations assessed on the index. It is also useful in assessing which areas need to be improved in order for a country to attain a higher standard of living.

Indicators that make up the index include the life expectancy of a person born in that country, their expected and mean years of schooling and their Gross National Income (GNI) per capita. Through a complex calculation, the indicators are combined to give a HDI value, measuring the standard of living in the respective country. The higher the life expectancy, the number of expected and the number of mean years at school and the GNI per capita of a country’s populace would generally mean a higher HDI value, meaning that respective country has a higher standard of living. A HDI value closer to 1 means the country has a high standard of living, while a HDI value closer to 0 means the country has a lower standard of living.

The dataset used contains the HDI values and ranks for countries in 2017, in addition to four columns listing the results of the four indicators used to measure the HDI value for each country. There are also two additional columns at the end of the dataset – a column that shows the results of the HDI rank subtracted from the GNI per capita rank and a column that shows the HDI rank for 2016.

The HDI index proves to be a fascinating dataset to use, because it raises several opportunities to ask questions about how the various indicators affect the living standards in various countries.  One interesting question that arises is which indicator is the best at predicting the HDI rank of each country? Does life expectancy, expected years of schooling, mean years of schooling or GNI per capita correlate the best with the rank of the HDI value for each country? 

The question is significant as with the answer, one could use it to predict the HDI rank of a country using just the one indicator that proves to be the best correlated with the ranking of countries based on their HDI ranks. This can give one a good idea of the living standards of a country without the need to know all the indicators or to have to work the calculation of all indicators that would attain the HDI value in the first place. It could also reveal on which indicator the HDI relies on the most in calculating its score for a given country.



## Methods and Results

To begin with, the appropriate packages from R have to be loaded; `rvest` and `dplyr`.

In [2]:
library(rvest)

Loading required package: xml2


In [3]:
library(dplyr)


Attaching package: 'dplyr'

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union



In [99]:
vignette('selectorgadget')

After loading the packages, the `read_html` function is used to read the website containing the data. The website is http://hdr.undp.org/en/composite/HDI. The function is saved as the object 'hweb'.

In [4]:
hweb = read_html("http://hdr.undp.org/en/composite/HDI")

In [102]:
?html_nodes

In [120]:
?html_table

Using the `html_nodes` function, the appropriate node for the dataset on the website is searched for. The dataset is a large table listing the HDI ranks, values and components for several countries.

In [5]:
html_nodes(hweb, 'table')

{xml_nodeset (1)}
[1] <table border="0" cellpadding="0" cellspacing="0" width="2268" style="bor ...

The data is scraped from the website onto this notebook. Because there were an irregular number of columns, for the function `html_table`, the variable `fill` had to be set at 'true'.

In [6]:
hweb2 = html_nodes(hweb, 'table') %>% .[[1]] %>% html_table(fill = TRUE)
hweb2

X1       X2                                                    
1            Table 1. Human Development Index and its components   
2                                                                  
3                                                                  
4   HDI rank Country                                               
5                                                                  
6            VERY HIGH HUMAN DEVELOPMENT                           
7   1        Norway                                                
8   2        Switzerland                                           
9   3        Australia                                             
10  4        Ireland                                               
11  5        Germany                                               
12  6        Iceland                                               
13  7        Hong Kong, China (SAR)                                
14  7        Sweden                                                
15  9        Singapore                                             
16  10       Netherlands                                           
17  11       Denmark                                               
18  12       Canada                                                
19  13       United States                                         
20  14       United Kingdom                                        
21  15       Finland                                               
22  16       New Zealand                                           
23  17       Belgium                                               
24  17       Liechtenstein                                         
25  19       Japan                                                 
26  20       Austria                                               
27  21       Luxembourg                                            
28  22       Israel                                                
29  22       Korea (Republic of)                                   
30  24       France                                                
... ...      ...                                                   
200 ..       Korea (Democratic People's Rep. of)                   
201 ..       Monaco                                                
202 ..       Nauru                                                 
203 ..       San Marino                                            
204 ..       Somalia                                               
205 ..       Tuvalu                                                
206                                                                
207          Human development groups                              
208          Very high human development                           
209          High human development                                
210          Medium human development                              
211          Low human development                                 
212                                                                
213          Developing countries                                  
214                                                                
215          Regions                                               
216          Arab States                                           
217          East Asia and the Pacific                             
218          Europe and Central Asia                               
219          Latin America and the Caribbean                       
220          South Asia                                            
221          Sub-Saharan Africa                                    
222                                                                
223          Least developed countries                             
224          Small island developing states                        
225                                                                
226          Organisation for Economic Co-operation and Deve

The dataset scraped from the website is a mess. To make things clearer, the `filter` function was used to clear all rows with blank space in the first column (X1), plus all countries without a defined HDI rank or value (any space filled with two or three periods).

In [31]:
hweb2 = filter(hweb2, X1!='', X1!='...', X1!='..')
hweb2

X1       X2                                 X3    X4  X5      X6  X7     
1   HDI rank Country                            Value     (years)     (years)
2   1        Norway                             0.953     82.3        17.9   
3   2        Switzerland                        0.944     83.5        16.2   
4   3        Australia                          0.939     83.1        22.9   
5   4        Ireland                            0.938     81.6        19.6   
6   5        Germany                            0.936     81.2        17.0   
7   6        Iceland                            0.935     82.9        19.3   
8   7        Hong Kong, China (SAR)             0.933     84.1        16.3   
9   7        Sweden                             0.933     82.6        17.6   
10  9        Singapore                          0.932     83.2        16.2   
11  10       Netherlands                        0.931     82.0        18.0   
12  11       Denmark                            0.929     80.9        19.1   
13  12       Canada                             0.926     82.5        16.4   
14  13       United States                      0.924     79.5        16.5   
15  14       United Kingdom                     0.922     81.7        17.4   
16  15       Finland                            0.920     81.5        17.6   
17  16       New Zealand                        0.917     82.0        18.9   
18  17       Belgium                            0.916     81.3        19.8   
19  17       Liechtenstein                      0.916     80.4    g   14.7   
20  19       Japan                              0.909     83.9        15.2   
21  20       Austria                            0.908     81.8        16.1   
22  21       Luxembourg                         0.904     82.0        14.0   
23  22       Israel                             0.903     82.7        15.9   
24  22       Korea (Republic of)                0.903     82.4        16.5   
25  24       France                             0.901     82.7        16.4   
26  25       Slovenia                           0.896     81.1        17.2   
27  26       Spain                              0.891     83.3        17.9   
28  27       Czechia                            0.888     78.9        16.9   
29  28       Italy                              0.880     83.2        16.3   
30  29       Malta                              0.878     81.0        15.9   
... ...      ...                                ...   ... ...     ... ...    
161 159      Mauritania                         0.520     63.4        8.6    
162 161      Madagascar                         0.519     66.3        10.6   
163 162      Uganda                             0.516     60.2        11.6   
164 163      Benin                              0.515     61.2        12.6   
165 164      Senegal                            0.505     67.5        9.7    
166 165      Comoros                            0.503     63.9        11.2   
167 165      Togo                               0.503     60.5        12.4   
168 167      Sudan                              0.502     64.7        7.4    
169 168      Afghanistan                        0.498     64.0        10.4   
170 168      Haiti                              0.498     63.6        9.3    
171 170      Côte d'Ivoire                      0.492     54.1        9.0    
172 171      Malawi                             0.477     63.7        10.8   
173 172      Djibouti                           0.476     62.6        6.2    
174 173      Ethiopia                           0.463     65.9        8.5    
175 174      Gambia                             0.460     61.4        9.2    
176 175      Guinea                             0.459     60.6        9.1    
177 176      Congo (Democratic Republic of the) 0.457     60.0        9.8    
178 177      Guinea-Bissau                      0.455     57.8        10.5   
179 178      Yemen                              0.452     65.2        9.0    
180 179      Eritrea      

This looks slightly better but it is still a mess. The headings have been cut off - however, that can be fixed later on.
Next, any column with no useful information in it (for example, if they are blank or are one of the many containin 'NA' repeated over and over) is selected out using a combination of the `subset` and `select` functions.
Note that under the `select` function, the `c()` function is listing the columns with blanks or 'NA's'. There is a dash before the `c()` function, indicating these are the columns that **are not** wanted.

In [32]:
hweb3 = subset(hweb2, select=-c(X4,X6,X8,X10,X12,X14,X16:X31))
hweb3

X1       X2                                 X3    X5      X7      X9     
1   HDI rank Country                            Value (years) (years) (years)
2   1        Norway                             0.953 82.3    17.9    12.6   
3   2        Switzerland                        0.944 83.5    16.2    13.4   
4   3        Australia                          0.939 83.1    22.9    12.9   
5   4        Ireland                            0.938 81.6    19.6    12.5   
6   5        Germany                            0.936 81.2    17.0    14.1   
7   6        Iceland                            0.935 82.9    19.3    12.4   
8   7        Hong Kong, China (SAR)             0.933 84.1    16.3    12.0   
9   7        Sweden                             0.933 82.6    17.6    12.4   
10  9        Singapore                          0.932 83.2    16.2    11.5   
11  10       Netherlands                        0.931 82.0    18.0    12.2   
12  11       Denmark                            0.929 80.9    19.1    12.6   
13  12       Canada                             0.926 82.5    16.4    13.3   
14  13       United States                      0.924 79.5    16.5    13.4   
15  14       United Kingdom                     0.922 81.7    17.4    12.9   
16  15       Finland                            0.920 81.5    17.6    12.4   
17  16       New Zealand                        0.917 82.0    18.9    12.5   
18  17       Belgium                            0.916 81.3    19.8    11.8   
19  17       Liechtenstein                      0.916 80.4    14.7    12.5   
20  19       Japan                              0.909 83.9    15.2    12.8   
21  20       Austria                            0.908 81.8    16.1    12.1   
22  21       Luxembourg                         0.904 82.0    14.0    12.1   
23  22       Israel                             0.903 82.7    15.9    13.0   
24  22       Korea (Republic of)                0.903 82.4    16.5    12.1   
25  24       France                             0.901 82.7    16.4    11.5   
26  25       Slovenia                           0.896 81.1    17.2    12.2   
27  26       Spain                              0.891 83.3    17.9    9.8    
28  27       Czechia                            0.888 78.9    16.9    12.7   
29  28       Italy                              0.880 83.2    16.3    10.2   
30  29       Malta                              0.878 81.0    15.9    11.3   
... ...      ...                                ...   ...     ...     ...    
161 159      Mauritania                         0.520 63.4    8.6     4.5    
162 161      Madagascar                         0.519 66.3    10.6    6.1    
163 162      Uganda                             0.516 60.2    11.6    6.1    
164 163      Benin                              0.515 61.2    12.6    3.6    
165 164      Senegal                            0.505 67.5    9.7     3.0    
166 165      Comoros                            0.503 63.9    11.2    4.8    
167 165      Togo                               0.503 60.5    12.4    4.8    
168 167      Sudan                              0.502 64.7    7.4     3.7    
169 168      Afghanistan                        0.498 64.0    10.4    3.8    
170 168      Haiti                              0.498 63.6    9.3     5.3    
171 170      Côte d'Ivoire                      0.492 54.1    9.0     5.2    
172 171      Malawi                             0.477 63.7    10.8    4.5    
173 172      Djibouti                           0.476 62.6    6.2     4.1    
174 173      Ethiopia                           0.463 65.9    8.5     2.7    
175 174      Gambia                             0.460 61.4    9.2     3.5    
176 175      Guinea                             0.459 60.6    9.1     2.6    
177 176      Congo (Democratic Republic of the) 0.457 60.0    9.8     6.8    
178 177      Guinea-Bissau                      0.455 57.8    10.5    3.0    
179 178      Yemen                              0.452 65.2    9.0     3.0    
180 179      Eritrea      

Slowly, the dataset is being cleared up. 

The first row of the dataset is actually made up  of parts of the headings that the dataset is supposed to had. This row needs to be filtered out and the current column headings (X1, X2 etc) need to be renamed.
First, the top row is filtered out. This dones by filtering out the row in column 'X2' with the word "Country" in it.

In [33]:
hweb3 = filter(hweb3, X2!="Country")
hweb3

X1  X2                                 X3    X5   X7   X9   X11    X13 X15
1   1   Norway                             0.953 82.3 17.9 12.6 68,012 5   1  
2   2   Switzerland                        0.944 83.5 16.2 13.4 57,625 8   2  
3   3   Australia                          0.939 83.1 22.9 12.9 43,560 18  3  
4   4   Ireland                            0.938 81.6 19.6 12.5 53,754 8   4  
5   5   Germany                            0.936 81.2 17.0 14.1 46,136 13  4  
6   6   Iceland                            0.935 82.9 19.3 12.4 45,810 13  6  
7   7   Hong Kong, China (SAR)             0.933 84.1 16.3 12.0 58,420 2   8  
8   7   Sweden                             0.933 82.6 17.6 12.4 47,766 9   7  
9   9   Singapore                          0.932 83.2 16.2 11.5 82,503 -6  8  
10  10  Netherlands                        0.931 82.0 18.0 12.2 47,900 5   10 
11  11  Denmark                            0.929 80.9 19.1 12.6 47,918 3   10 
12  12  Canada                             0.926 82.5 16.4 13.3 43,433 10  12 
13  13  United States                      0.924 79.5 16.5 13.4 54,941 -2  12 
14  14  United Kingdom                     0.922 81.7 17.4 12.9 39,116 13  14 
15  15  Finland                            0.920 81.5 17.6 12.4 41,002 10  15 
16  16  New Zealand                        0.917 82.0 18.9 12.5 33,970 18  16 
17  17  Belgium                            0.916 81.3 19.8 11.8 42,156 6   16 
18  17  Liechtenstein                      0.916 80.4 14.7 12.5 97,336 -15 16 
19  19  Japan                              0.909 83.9 15.2 12.8 38,986 9   19 
20  20  Austria                            0.908 81.8 16.1 12.1 45,415 0   20 
21  21  Luxembourg                         0.904 82.0 14.0 12.1 65,016 -13 26 
22  22  Israel                             0.903 82.7 15.9 13.0 32,711 13  21 
23  22  Korea (Republic of)                0.903 82.4 16.5 12.1 35,945 8   22 
24  24  France                             0.901 82.7 16.4 11.5 39,254 2   23 
25  25  Slovenia                           0.896 81.1 17.2 12.2 30,594 12  24 
26  26  Spain                              0.891 83.3 17.9 9.8  34,258 7   25 
27  27  Czechia                            0.888 78.9 16.9 12.7 30,588 11  27 
28  28  Italy                              0.880 83.2 16.3 10.2 35,299 3   28 
29  29  Malta                              0.878 81.0 15.9 11.3 34,396 3   29 
30  30  Estonia                            0.871 77.7 16.1 12.7 28,993 10  30 
... ... ...                                ...   ...  ...  ...  ...    ... ...
160 159 Mauritania                         0.520 63.4 8.6  4.5  3,592  -13 159
161 161 Madagascar                         0.519 66.3 10.6 6.1  1,358  20  158
162 162 Uganda                             0.516 60.2 11.6 6.1  1,658  13  162
163 163 Benin                              0.515 61.2 12.6 3.6  2,061  2   161
164 164 Senegal                            0.505 67.5 9.7  3.0  2,384  -2  165
165 165 Comoros                            0.503 63.9 11.2 4.8  1,399  15  163
166 165 Togo                               0.503 60.5 12.4 4.8  1,453  14  164
167 167 Sudan                              0.502 64.7 7.4  3.7  4,119  -25 165
168 168 Afghanistan                        0.498 64.0 10.4 3.8  1,824  0   168
169 168 Haiti                              0.498 63.6 9.3  5.3  1,665  6   167
170 170 Côte d'Ivoire                      0.492 54.1 9.0  5.2  3,481  -22 169
171 171 Malawi                             0.477 63.7 10.8 4.5  1,064  14  170
172 172 Djibouti                           0.476 62.6 6.2  4.1  3,392  -21 170
173 173 Ethiopia                           0.463 65.9 8.5  2.7  1,719  -1  173
174 174 Gambia                             0.460 61.4 9.2  3.5  1,516  4   173
175 175 Guinea                             0.459 60.6 9.1  2.6  2,067  -11 177
176 176 Congo (Democratic Republic of the) 0.457 60.0 9.8  6.8  796    12  176
177 177 Guinea-Bissau                      0.455 57.8 10.5 3.0  1,552  0   175
178 178 Yemen                              0.452 65.2 

Now the columns are renamed. It should be noted that some of the columns are not named as they were in the original dataset due to concerns over space. However, most of the changes were three or four words from some columns explaining what 'HDI' and 'GNI' mean. For reference, 'HDI' means 'Human Development Index' and 'GNI' means 'Gross National Income'. Brackets have been eliminated because the R program kept mistaking the use of brackets in the column title as part of a function. Also, underscores have been placed between the words in each column name to connect them up for when data manupulation occurs later. 

In [41]:
colnames(hweb3) = c("HDI_rank_2017", "Country", "HDI_Value", "Life_expectancy_at_birth_in_years", "Expected_years_of_schooling_in_years",
                    "Mean_years_of_schooling_in_years", "GNI_per_capita_in_USdollars", "GNI_per_capita_rank_-_HDI_rank", "HDI_rank_2016")
hweb3

HDI_rank_2017 Country                            HDI_Value
1   1             Norway                             0.953    
2   2             Switzerland                        0.944    
3   3             Australia                          0.939    
4   4             Ireland                            0.938    
5   5             Germany                            0.936    
6   6             Iceland                            0.935    
7   7             Hong Kong, China (SAR)             0.933    
8   7             Sweden                             0.933    
9   9             Singapore                          0.932    
10  10            Netherlands                        0.931    
11  11            Denmark                            0.929    
12  12            Canada                             0.926    
13  13            United States                      0.924    
14  14            United Kingdom                     0.922    
15  15            Finland                            0.920    
16  16            New Zealand                        0.917    
17  17            Belgium                            0.916    
18  17            Liechtenstein                      0.916    
19  19            Japan                              0.909    
20  20            Austria                            0.908    
21  21            Luxembourg                         0.904    
22  22            Israel                             0.903    
23  22            Korea (Republic of)                0.903    
24  24            France                             0.901    
25  25            Slovenia                           0.896    
26  26            Spain                              0.891    
27  27            Czechia                            0.888    
28  28            Italy                              0.880    
29  29            Malta                              0.878    
30  30            Estonia                            0.871    
... ...           ...                                ...      
160 159           Mauritania                         0.520    
161 161           Madagascar                         0.519    
162 162           Uganda                             0.516    
163 163           Benin                              0.515    
164 164           Senegal                            0.505    
165 165           Comoros                            0.503    
166 165           Togo                               0.503    
167 167           Sudan                              0.502    
168 168           Afghanistan                        0.498    
169 168           Haiti                              0.498    
170 170           Côte d'Ivoire                      0.492    
171 171           Malawi                             0.477    
172 172           Djibouti                           0.476    
173 173           Ethiopia                           0.463    
174 174           Gambia                             0.460    
175 175           Guinea                             0.459    
176 176           Congo (Democratic Republic of the) 0.457    
177 177           Guinea-Bissau                      0.455    
178 178           Yemen                              0.452    
179 179           Eritrea                            0.440    
180 180           Mozambique                         0.437    
181 181           Liberia                            0.435    
182 182           Mali                               0.427    
183 183           Burkina Faso                       0.423    
184 184           Sierra Leone                       0.419    
185 185           Burundi                            0.417    
186 186           Chad                               0.404    
187 187           South Sudan                        0.388    
188 188           Central African Republic           0.367    
189 189           Niger                              0.354    
    Life_expectancy_at_birth_in_years Expected_years_of_schooling_in_years
1   82.3               

Now the data is ready to be manipulated.

To repeat, we want to find out which indicator (life expectancy, expected years of schooling, mean years of schooling or GNI per capita) is the closest correlated with the HDI rank of each country. The plan to calculate this is to give a rank for each value in each indicator and use to determine a correlation coefficient between the ranks of the indicator and the HDI ranks. A correlation coefficient measures the correlation between two or more datasets. A coefficient closer to -1 is a strong negative correlation, while a coefficient closer to +1 is a positive correlation. A coefficient around 0 shows no or weak correlation.

For example, the data for 'Life_expectancy_at_birth' column shall be manpulated first.

In [68]:
?rank

Warning message in file(out, "wt"):
"cannot open file 'C:\Users\User\AppData\Local\Temp\Rtmp8mYeO8\file18cc711969bb': No such file or directory"Warning message in file(con, "r"):
"cannot open file 'C:\Users\User\AppData\Local\Temp\Rtmp8mYeO8\file18cc711969bb': No such file or directory"ERROR while rich displaying an object: Error in file(con, "r"): cannot open the connection

Traceback:
1. FUN(X[[i]], ...)
2. tryCatch(withCallingHandlers({
 .     rpr <- mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler), error = outer_handler)
3. tryCatchList(expr, classes, parentenv, handlers)
4. tryCatchOne(expr, names, parentenv, handlers[[1L]])
5. doTryCatch(return(expr), name, parentenv, handler)
6. withCallingHandlers({
 .     rpr <- mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler)
7. mime2repr[[mime]](obj)
8. rep

First, an object for the 'Life_expectancy_at_birth' data is created to keep things simple.

In [57]:
le = hweb3$Life_expectancy_at_birth_in_years
le

[1] "82.3" "83.5" "83.1" "81.6" "81.2" "82.9" "84.1" "82.6" "83.2" "82.0"
 [11] "80.9" "82.5" "79.5" "81.7" "81.5" "82.0" "81.3" "80.4" "83.9" "81.8"
 [21] "82.0" "82.7" "82.4" "82.7" "81.1" "83.3" "78.9" "83.2" "81.0" "77.7"
 [31] "81.4" "80.7" "77.8" "77.4" "81.7" "74.8" "78.3" "77.0" "77.4" "74.7"
 [41] "74.7" "81.4" "77.0" "79.7" "76.1" "77.8" "76.7" "77.3" "71.2" "77.3"
 [51] "74.9" "75.6" "73.1" "75.8" "77.6" "74.8" "75.5" "76.1" "70.0" "76.2"
 [61] "73.4" "73.7" "80.0" "76.0" "74.9" "78.2" "75.3" "78.5" "70.8" "76.5"
 [71] "73.4" "74.4" "79.9" "77.3" "73.8" "75.5" "77.1" "74.7" "75.7" "72.1"
 [81] "79.8" "75.9" "74.8" "75.5" "76.3" "76.4" "76.6" "72.1" "75.2" "74.6"
 [91] "75.7" "70.4" "69.5" "74.0" "74.5" "75.9" "76.1" "73.2" "73.3" "71.5"
[101] "67.6" "77.6" "78.0" "75.2" "71.4" "70.6" "73.6" "72.1" "68.0" "66.5"
[111] "73.2" "71.7" "69.2" "63.4" "71.7" "69.4" "76.5" "69.5" "73.6" "70.0"
[121] "73.8" "71.1" "76.1" "75.7" "73.0" "66.8" "73.7" "71.2" "64.9" "68.8"
[131] "69.3" "69.2" "73.8" "70.6" "66.5" "72.8" "65.1" "72.3" "67.0" "63.0"
[141] "57.9" "67.3" "66.8" "58.3" "62.3" "69.3" "61.8" "66.7" "70.6" "66.6"
[151] "58.6" "71.0" "65.7" "66.3" "71.0" "61.7" "53.9" "67.5" "54.6" "63.4"
[161] "66.3" "60.2" "61.2" "67.5" "63.9" "60.5" "64.7" "64.0" "63.6" "54.1"
[171] "63.7" "62.6" "65.9" "61.4" "60.6" "60.0" "57.8" "65.2" "65.5" "58.9"
[181] "63.0" "58.5" "60.8" "52.2" "57.9" "53.2" "57.3" "52.9" "60.4"

   Then, each value for life expectancy is given a rank. `rank(desc())` is used because the ranking should be done in descending order - the largest value for life expectancy should have the lowest rank. The function `ties.method` denotes what rank the ties between values should be given. "min" means ties should be given the smallest rank - as is with the HDI ranks.

In [97]:
lerank = rank(desc(le), ties.method = "min")
lerank

[1]  14   3   7  21  26   8   1  11   5  15  29  12  36  19  22  15  25  31
 [19]   2  18  15   9  13   9  27   4  37   5  28  44  23  30  42  47  19  82
 [37]  39  53  47  85  85  23  53  35  62  42  55  49 115  49  80  73 104  69
 [55]  45  82  74  62 125  61  99  95  32  66  80  40  77  38 120  57  99  90
 [73]  33  49  92  74  52  85  70 108  34  67  82  74  60  59  56 108  78  88
 [91]  70 124 127  91  89  67  62 102 101 113 136  45  41  78 114 121  97 108
[109] 135 145 102 111 132 160 111 129  57 127  97 125  92 117  62  70 105 141
[127]  95 115 154 134 130 132  92 121 145 106 153 107 140 162 180 139 141 179
[145] 165 130 166 143 121 144 177 118 150 147 118 167 186 137 184 160 147 174
[163] 169 137 157 172 155 156 159 185 158 164 149 168 171 175 182 152 151 176
[181] 162 178 170 189 180 187 183 188 173

To make things simpler, 'HDI_rank_2017' is turned into an object.

In [98]:
hdirank = hweb3$HDI_rank_2017
hdirank

[1] "1"   "2"   "3"   "4"   "5"   "6"   "7"   "7"   "9"   "10"  "11"  "12" 
 [13] "13"  "14"  "15"  "16"  "17"  "17"  "19"  "20"  "21"  "22"  "22"  "24" 
 [25] "25"  "26"  "27"  "28"  "29"  "30"  "31"  "32"  "33"  "34"  "35"  "35" 
 [37] "37"  "38"  "39"  "39"  "41"  "41"  "43"  "44"  "45"  "46"  "47"  "48" 
 [49] "49"  "50"  "51"  "52"  "53"  "54"  "55"  "56"  "57"  "58"  "58"  "60" 
 [61] "60"  "62"  "63"  "64"  "65"  "66"  "67"  "68"  "69"  "70"  "70"  "72" 
 [73] "73"  "74"  "75"  "76"  "77"  "78"  "79"  "80"  "80"  "80"  "83"  "83" 
 [85] "85"  "86"  "86"  "88"  "89"  "90"  "90"  "92"  "92"  "94"  "95"  "95" 
 [97] "97"  "98"  "99"  "100" "101" "101" "103" "104" "105" "106" "106" "108"
[109] "108" "110" "110" "112" "113" "113" "115" "116" "116" "118" "119" "120"
[121] "121" "122" "123" "124" "125" "125" "127" "127" "129" "130" "131" "132"
[133] "133" "134" "134" "136" "137" "138" "139" "140" "141" "142" "143" "144"
[145] "144" "146" "147" "148" "149" "150" "151" "152" "153" "154" "155" "156"
[157] "157" "158" "159" "159" "161" "162" "163" "164" "165" "165" "167" "168"
[169] "168" "170" "171" "172" "173" "174" "175" "176" "177" "178" "179" "180"
[181] "181" "182" "183" "184" "185" "186" "187" "188" "189"

The inverted commas around the values for the HDI rank indicate that these values are not numeric. They must be converted to numeric values first.

In [99]:
hdirank = as.numeric(hdirank)

Now, using the `cor()` function, the correlation between HDI ranks and life expectancy can be determined.

In [161]:
cor(lerank, hdirank)

[1] 0.9134554

0.91345536617642 shows a strong correlation between the HDI ranks and life expectancy ranks. However, the other indicators may have stronger correlations.

First, for the 'GNI_per_capita_in_USdollars' column, we have to remove the commas from the values within it. We use `gsub` to tell R that the commas(",") should be removed and replaced with nothing (""). We then change the values to a numerical value using `as.numeric`.

In [ ]:
?gsub

In [150]:
class(hweb3$GNI_per_capita_in_USdollars)
gsub(",", "", hweb3$GNI_per_capita_in_USdollars)
gn = as.numeric(gsub(",", "", hweb3$GNI_per_capita_in_USdollars))
gn

[1] "character"

[1] "68012"  "57625"  "43560"  "53754"  "46136"  "45810"  "58420"  "47766" 
  [9] "82503"  "47900"  "47918"  "43433"  "54941"  "39116"  "41002"  "33970" 
 [17] "42156"  "97336"  "38986"  "45415"  "65016"  "32711"  "35945"  "39254" 
 [25] "30594"  "34258"  "30588"  "35299"  "34396"  "28993"  "24648"  "31568" 
 [33] "26150"  "67805"  "47574"  "28314"  "116818" "29467"  "76427"  "49680" 
 [41] "25002"  "27315"  "41580"  "21910"  "25393"  "22162"  "18461"  "36290" 
 [49] "24233"  "16779"  "18740"  "22646"  "16323"  "26681"  "19930"  "70524" 
 [57] "26107"  "15843"  "22626"  "19130"  "12831"  "26077"  "14636"  "24804" 
 [65] "20189"  "19178"  "13019"  "11886"  "28622"  "20764"  "9186"   "23978" 
 [73] "7524"   "16944"  "12864"  "11326"  "11716"  "10672"  "13755"  "15600" 
 [81] "13378"  "12505"  "9144"   "15516"  "13802"  "15270"  "10347"  "8130"  
 [89] "11789"  "12938"  "11695"  "8324"   "10103"  "13921"  "8288"   "10275" 
 [97] "7846"   "5547"   "10499"  "13306"  "15534"  "13567"  "8344"   "5909"  
[105] "6470"   "7166"   "5125"   "11100"  "15594"  "16431"  "8380"   "5554"  
[113] "9154"   "11923"  "10355"  "10846"  "5859"   "6714"   "5055"   "17789" 
[121] "6868"   "3255"   "7340"   "5157"   "5983"   "7447"   "7278"   "3317"  
[129] "9387"   "6353"   "3843"   "6846"   "4215"   "8065"   "3042"   "3677"  
[137] "5694"   "2995"   "6070"   "4096"   "19513"  "2961"   "2941"   "7620"  
[145] "3557"   "3413"   "5790"   "5567"   "2471"   "5311"   "3315"   "1872"  
[153] "3403"   "2655"   "2337"   "1683"   "5231"   "1811"   "3255"   "3592"  
[161] "1358"   "1658"   "2061"   "2384"   "1399"   "1453"   "4119"   "1824"  
[169] "1665"   "3481"   "1064"   "3392"   "1719"   "1516"   "2067"   "796"   
[177] "1552"   "1239"   "1750"   "1093"   "667"    "1953"   "1650"   "1240"  
[185] "702"    "1750"   "963"    "663"    "906"

[1]  68012  57625  43560  53754  46136  45810  58420  47766  82503  47900
 [11]  47918  43433  54941  39116  41002  33970  42156  97336  38986  45415
 [21]  65016  32711  35945  39254  30594  34258  30588  35299  34396  28993
 [31]  24648  31568  26150  67805  47574  28314 116818  29467  76427  49680
 [41]  25002  27315  41580  21910  25393  22162  18461  36290  24233  16779
 [51]  18740  22646  16323  26681  19930  70524  26107  15843  22626  19130
 [61]  12831  26077  14636  24804  20189  19178  13019  11886  28622  20764
 [71]   9186  23978   7524  16944  12864  11326  11716  10672  13755  15600
 [81]  13378  12505   9144  15516  13802  15270  10347   8130  11789  12938
 [91]  11695   8324  10103  13921   8288  10275   7846   5547  10499  13306
[101]  15534  13567   8344   5909   6470   7166   5125  11100  15594  16431
[111]   8380   5554   9154  11923  10355  10846   5859   6714   5055  17789
[121]   6868   3255   7340   5157   5983   7447   7278   3317   9387   6353
[131]   3843   6846   4215   8065   3042   3677   5694   2995   6070   4096
[141]  19513   2961   2941   7620   3557   3413   5790   5567   2471   5311
[151]   3315   1872   3403   2655   2337   1683   5231   1811   3255   3592
[161]   1358   1658   2061   2384   1399   1453   4119   1824   1665   3481
[171]   1064   3392   1719   1516   2067    796   1552   1239   1750   1093
[181]    667   1953   1650   1240    702   1750    963    663    906

Finally, the 'GNI_per_capita_in_USdollars' column values have been sorted.

One may ask why 'GNI_per_capita_rank_-_HDI_rank' was not used instead. Surely it would have been easier to just sum that particular column up, instead of having to code extra to turn the 'GNI_per_capita' column values into numerical values. However, there are two reasons as to why the latter option was taken. One was because the GNI per capita rank is supposed to be subtracted *from* the HDI rank, of which the 'GNI_per_capita_rank_-_HDI_rank does the opposite. The second reason is because it is better to use fresh data than to rely on secondary sources. 

Now that the 'GNI_per_capita_in_USdollars' column values have been sorted, the other two indicators, expected years of schooling and mean years of schooling, need to be turned to objects.

In [130]:
es = hweb3$Expected_years_of_schooling_in_years
ms = hweb3$Mean_years_of_schooling_in_years

In [162]:
?cor

Warning message in file(out, "wt"):
"cannot open file 'C:\Users\User\AppData\Local\Temp\Rtmp8mYeO8\file18cc7a616b1a': No such file or directory"Warning message in file(con, "r"):
"cannot open file 'C:\Users\User\AppData\Local\Temp\Rtmp8mYeO8\file18cc7a616b1a': No such file or directory"ERROR while rich displaying an object: Error in file(con, "r"): cannot open the connection

Traceback:
1. FUN(X[[i]], ...)
2. tryCatch(withCallingHandlers({
 .     rpr <- mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler), error = outer_handler)
3. tryCatchList(expr, classes, parentenv, handlers)
4. tryCatchOne(expr, names, parentenv, handlers[[1L]])
5. doTryCatch(return(expr), name, parentenv, handler)
6. withCallingHandlers({
 .     rpr <- mime2repr[[mime]](obj)
 .     if (is.null(rpr)) 
 .         return(NULL)
 .     prepare_content(is.raw(rpr), rpr)
 . }, error = error_handler)
7. mime2repr[[mime]](obj)
8. rep

The objects containing the values for each indicator are then ranked.

In [163]:
esrank = rank(desc(es), ties.method = "min")
msrank = rank(desc(ms), ties.method = "min")
gnrank = rank(desc(gn), ties.method = "min")

In [164]:
cor(esrank, hdirank)
cor(msrank, hdirank)
cor(gnrank, hdirank)


[1] 0.3042133

[1] -0.2386791

[1] 0.9517759

The correlation coefficients between the HDI ranks and each indicator are as follows:
- the correlation coefficient of the hdi ranks and life expectancy ranks is 0.913455366176421.
- the correlation coefficient of the hdi ranks and mean schooling years' ranks is 0.304213266854488.
- the correlation coefficient of the hdi ranks and mean schooling years' ranks is -0.238679106739706.
- the correlation coefficient of the hdi ranks and the GNI per capita ranks is 0.951775940760708.

## Discussion

To recap, the purpose of this assignment was to find out which of the four indicators making up the Human Development Index is the closest correlated with the HDI rank of each country. A rank was given to each value in each indicator and a correlation coefficient was determined between the ranks of the indicator and the ranks of the HDI values. The indicator with the correlation coefficient closest to 1 is the indicator most positively correlated with the HDI rank. Because the ‘GNI_per_capita_in_USdollars’ indicator has the coefficient closest to 1, with a (rounded up) coefficient of 0.952, it provides the best indicator for predicting what the HDI rank will be.

This does not mean ‘GNI per capita’ is the perfect method for predicting HDI values or ranks or that one should only use ‘GNI per capita’ when assessing the living standards of a country. The most likely explanation for the results is that the calculations used to work out the HDI values put a lot more weight on ‘GNI per capita’ than on the other three indicators. Effectively, one could suggest that these results point out a flaw in the HDI calculations and where it can be improved in order to give equal weight to all indicators. 

According to the Human Development Report’s [website](http://hdr.undp.org/en/faq-page/human-development-index-hdi#t292n287), from which the dataset was obtained, GNI per capita only reflects the average national income and not as to whether this translates to better health, education or other improvements in quality of life. However, this does not explain as to why there is such a strong correlation between GNI per capita and HDI. One interpretation is that a higher GNI per capita is actually a result of improved human development, rather than a cause or indicator of it, meaning a higher HDI score of a country will result in a higher GNI per capita.

However, a more [extensive study](https://www.tandfonline.com/doi/full/10.1080/1331677X.2017.138317) published in 2017 claimed that correlations between economic indicators and HDI are a lot more unclear than would be thought. The study claimed that there is a weak correlation between GDP (Gross Domestic Product) and HDI but a stronger correlation between GDP per capita and HDI. The study then points out that the distribution of wealth and resources is a better indicator of human development than the mere accumulation of them. While they relied more on GDP per capita in their study, the fact that GNI per capita is [based on GDP per capita](https://en.wikipedia.org/wiki/Gross_national_income) could explain the high correlation achieved in this assignment – more highly developed countries generally have a higher GNI per capita due to a better distribution of wealth.

To summarise, GNI per capita is the most highly correlated indicator of HDI to the HDI ranks. Among the potential reasons for this are statistical bias, it being an outcome of the HDI value or it being a perfectly valid factor indicating a better distribution of wealth and resources, hence leading to a better HDI. However, GNI per capita is only one indicator and should not be used alone in assessing the living standards of a country.

Help for working the correlation coefficient in R came from http://www.r-tutor.com/elementary-statistics/numerical-measures/correlation-coefficient.